src/lib/support/ScopedBuffer.h

```diff
@@ -0,0 +1,152 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Defines scoped auto-free buffers for CHIP.
+ *
+ */
+
+#ifndef CHIP_SCOPED_BUFFER_H_
+#define CHIP_SCOPED_BUFFER_H_
+
+#include <lib/support/CHIPMem.h>
+
+namespace chip {
+namespace Platform {
+
+namespace Impl {
+
+/**
+ * Represents a memory buffer that is auto-freed in the destructor.
+ */
+template <class Impl>
+class ScopedMemoryBufferBase
+{
+public:
+    ScopedMemoryBufferBase() {}
+    ScopedMemoryBufferBase(const ScopedMemoryBufferBase &) = delete;
+    ScopedMemoryBufferBase & operator=(const ScopedMemoryBufferBase & other) = delete;
+
+    ~ScopedMemoryBufferBase() { Free(); }
+
+    void * Ptr() { return mBuffer; }
+    const void * Ptr() const { return mBuffer; }
+
+    template <typename T>
```

Comment:
I wonder whether it makes sense to make the following changes:

1. Add a template parameter for the type of the things in the buffer (to either the base class or the underlying class).  Let's call that type `Type`.
1. `static_assert` that `Type` is a POD type, to avoid issues with people putting things with destructors (which will not get run) in the buffer
1. If we have static knowledge about alignment of the things our allocator returns maybe `static_assert` that `Type` does not have stronger alignment requirements.
1. Add conversion operators returning `Type*` and `const Type*`.
1. Maybe add an `operator[]` returning `Type&` (and a const equivalent).

This would still allow effectively doing `reinterpret_cast` via the `Ptr` method if we want to allow that, but would also allow "normal" use to specify the type up front and not have to mention it at every place where the buffer is actually being used.  The various places in the changeset that use `someBuffer.Ptr<sometype>()` could just use `someBuffer`.

Thoughts?

Reply:
I considered that however my intent is to avoid template explosion (this was the main objection I heard about using templates in embedded). I do not want duplicate implementations inside flash for uint8_t, char, etc.

In this approach, we have only the Get duplicated, but that one is a trivial function and inlined, so it should go away and we get 0 cost abstraction.

Reply:
We may refactor it later... maybe subclassing works without exploding things (just make cast operators easy to use). I would defer it for the future though. At first I thought that this Memory alloc/free is used all over in chip, but in reality we have very limited places where we actually use it.

Reply:
> maybe subclassing works without exploding things

Right.  Codesize explosion is a good reason to not template the base class.  Templating the derived class, and making sure that the templated thing only has inline methods, would avoid the codesize problems, though.

I'm probably ok with this change as a followup, especially since it would involve codesize measurements, but I do think we should make it, not just for ease of use but also because it enables us to add safety asserts (like the `is_pod` assertion I suggested).  Are you willing to take the time to do this?

Reply:
I can do it, please add a PR and assign it to me.
I am unsure timing wise, however it seems small enough of a change that I can do it.
Also thinking we probably want a implementation for 'secure memory' to call a clear on every free. Our existing code seems somewhat buggy in that regard.

Reply:
I agree with Boris, the current implementation strikes me as type-unsafe and we should be able to fix it without increasing code size.

I would suggest removing the template parameter from Ptr<> in this initial PR, that's really just hiding a typecast under a less obvious syntax now. Potentially unsafe casts are the last thing we want to hide in a benign-looking accessor function.